rfc/rfc-79/rfc-90.md

```diff
@@ -0,0 +1,310 @@
+  executed (if execution of clustering is handled async) but that can be minimized by allowing clean to only do this if
+  instant is very old.
+- If a cancellable clustering plan was requested for cancellation but never fully aborted (again due to the corresponding
+  worker never running or repeatedly failing), `CLEAN` will take up task of cleaning up and aborting the inflight instant.
+
+Note that a cancellable clustering plan that has been requestd for cancellation can still be safely cleaned up immediately by execute_abort API - the goal here
+is just to make sure, an inflight plan won't stay on the timeline for an unbounded amount of time but also won't be
+likely to be prematurely cleaned up by clean before it has a chance to be executed.
+
+An additional config "cancellation-policy" will be added to the table service plan to indicate when it is eligible to be
+cancelled and aborted by `CLEAN` operation. This policy can be a threshold of hours or instants on timeline, where if
+that # of hours/instants have elapsed since the plan was scheduled(w/o being executed), any call to `CLEAN` operation
+will also cancel and abort the instant. This policy should be configured by the worker scheduling a cancellable table
+service plan, based on the amount of time they expect the plan to remain on the timeline before being picked up for
+execution. For example, if the plan is expected to have its execution deferred to a few hours later, then the
+cancellation-policy should be lenient in allowing the plan to remain many hours on the timeline before it could be
+deemed eligible for abortion. Note that this cancellation policy is not used to infer whether a clustering plan is
+currently being executed - similar to how concurrent ingestion writes are safely rolled back by lazy clean policy, A
+cancellable clustering plan can only be aborted once it is confirmed that an ongoing writer is no longer progressing it
+(which the aforementioned execute_abort API already enforces). Note that the cancellation-policy is only required to be honored by `CLEAN`
+
+To implement this optional feature, the clean operation can use both of these APIs in order to cancel any incomplete plans that have
+satisfied their cancellation policy or already been requested for cancellation. This is analagous to how clean already schedules and
+executes the rollback of any failed ingestion writes. Specifically, clean will now perform the following steps
+
+1. Iterate through each inflight cancellable instant, for each instant that has met their cancellation policy, call
+   request_cancel if not already has an request for cancellation.
+2. Iterate through each instant in /.cancel and call execute_abort. Note that the request_cancel and execute_abort APIs
+   are responsible for ensuring that already-committed instants aren't cancelled and that ongoing writers aren't
+   concurrently having their files cleaned up by a writer doing cancellation.
```

Comment:
CLEAN might see a plan that meets the time threshold but is actually running right now. The design states a plan that is truly in active execution is protected by the heartbeat. That logic should be carefully spelled out to avoid inadvertently cleaning an actually‐running plan.

If CLEAN sees a plan that meets the threshold, presumably it calls `request_cancel` and (if no heartbeat is found) calls `execute_abort` in quick succession. This effectively lumps the two actions into one. That is fine, but clarifying that the general approach is "try to request_cancel, then see if a heartbeat exists, if not, proceed to execute_abort" would help.

Reply:
Oh I didn't mention the heartbeating again here since execute_abort already implicitly checks for heartbeat existence as per the design. But I can mention that again here to clarify.